src/Features/LanguageServer/Protocol/Handler/Completion/CompletionHandler.cs

```diff
@@ -68,38 +97,112 @@ public CompletionHandler(ILspSolutionProvider solutionProvider) : base(solutionP
+            {
+                var completeDisplayText = item.DisplayTextPrefix + item.DisplayText + item.DisplayTextSuffix;
+                var completionItem = new TCompletionItem
                 {
-                    Label = item.DisplayTextPrefix + item.DisplayText + item.DisplayTextSuffix,
-                    InsertText = item.Properties.ContainsKey("InsertionText") ? item.Properties["InsertionText"] : item.DisplayText,
+                    Label = completeDisplayText,
+                    InsertText = item.Properties.ContainsKey("InsertionText") ? item.Properties["InsertionText"] : completeDisplayText,
                     SortText = item.SortText,
                     FilterText = item.FilterText,
                     Kind = GetCompletionKind(item.Tags),
-                    Data = new CompletionResolveData { TextDocument = request.TextDocument, Position = request.Position, DisplayText = item.DisplayText },
+                    Data = new CompletionResolveData
+                    {
+                        TextDocument = request.TextDocument,
+                        Position = request.Position,
+                        DisplayText = item.DisplayText,
+                        CompletionTrigger = completionTrigger,
+                    },
                     Preselect = item.Rules.SelectionBehavior == CompletionItemSelectionBehavior.HardSelection,
+                    CommitCharacters = GetCommitCharacters(item)
                 };
+
+                return completionItem;
+            }
+
+            static string[]? GetCommitCharacters(CompletionItem item)
+            {
+                var commitCharacterRules = item.Rules.CommitCharacterRules;
+
+                // If the item doesn't have any special rules, just use the default commit characters.
+                if (commitCharacterRules.IsEmpty)
+                {
+                    return null;
+                }
+
+                using var _ = PooledHashSet<char>.GetInstance(out var commitCharacters);
+                commitCharacters.AddAll(CompletionRules.Default.DefaultCommitCharacters);
+                foreach (var rule in commitCharacterRules)
+                {
+                    switch (rule.Kind)
+                    {
+                        case CharacterSetModificationKind.Add:
+                            commitCharacters.UnionWith(rule.Characters);
+                            continue;
+                        case CharacterSetModificationKind.Remove:
+                            commitCharacters.ExceptWith(rule.Characters);
+                            continue;
+                        case CharacterSetModificationKind.Replace:
+                            commitCharacters.Clear();
+                            commitCharacters.AddRange(rule.Characters);
+                            break;
+                    }
+                }
+
+                return commitCharacters.Select(c => c.ToString()).ToArray();
```

Comment:
is this called for every item.  seems like a lot of work/allocs.  can we cache the characters assocaited per `item.Rules.CommitCharacterRules` or `item.Rules` to avoid all this work/allocation?

Reply:
Added a cache per request - this avoids most of the work.
I didn't make the cache save between requests as it could grow infinitely and I didn't find an existing easy cache + eviction impl
(this is from the reverted PR)